File: ieee/ieee754.js
// Convert float to IEEE-754 binary string (32-bit)
function ieee754ToBin(num) {
    // Create a 4-byte ArrayBuffer and a DataView to access bits
    const buffer = new ArrayBuffer(4);
    const view = new DataView(buffer);

    // Write the float32 into the buffer
    view.setFloat32(0, num, false); // false = big-endian

    // Read as unsigned 32-bit integer
    const bits = view.getUint32(0, false);

    // Build binary string
    let binStr = "";
    for (let i = 31; i >= 0; i--) {
        binStr += ((bits >> i) & 1) ? "1" : "0";
    }
    return binStr;
}

// Convert IEEE-754 binary string to float
function binToIEEE754(binStr) {
    if (binStr.length !== 32) {
        throw new Error("Binary string must be 32 bits.");
    }

    let bits = 0;
    for (let i = 0; i < 32; i++) {
        if (binStr[i] === "1") {
            bits |= (1 << (31 - i)) >>> 0;
        } else if (binStr[i] !== "0") {
            throw new Error("Invalid character in binary string.");
        }
    }

    // Put bits into a buffer and read as float32
    const buffer = new ArrayBuffer(4);
    const view = new DataView(buffer);
    view.setUint32(0, bits, false);
    return view.getFloat32(0, false);
}

// Main-like function
function main() {
    const args = process.argv.slice(2);
    if (args.length !== 1) {
        console.log(`Usage: node ${process.argv[1]} <float_number>`);
        process.exit(1);
    }

    const input = parseFloat(args[0]);
    if (isNaN(input)) {
        console.error("Invalid float number.");
        process.exit(1);
    }

    // Convert float to binary
    const binary = ieee754ToBin(input);
    console.log(`IEEE-754 binary of ${input.toFixed(6)}: ${binary}`);

    // Convert binary back to float
    const recovered = binToIEEE754(binary);
    console.log(`Recovered float from binary: ${recovered.toFixed(6)}`);
}

// Run if executed directly
if (require.main === module) {
    main();
}

File: ieee/ieee754.py
import struct
import sys

# Convert float to IEEE-754 binary string (32-bit)
def ieee754_to_bin(num: float) -> str:
    # Pack float into 4 bytes (big-endian)
    packed = struct.pack('>f', num)
    # Unpack as unsigned 32-bit integer
    bits = struct.unpack('>I', packed)[0]
    # Format as 32-bit binary string
    return ''.join('1' if (bits >> i) & 1 else '0' for i in range(31, -1, -1))

# Convert IEEE-754 binary string to float
def bin_to_ieee754(bin_str: str) -> float:
    if len(bin_str) != 32:
        raise ValueError("Binary string must be 32 bits.")
    bits = 0
    for i, ch in enumerate(bin_str):
        if ch == '1':
            bits |= 1 << (31 - i)
        elif ch != '0':
            raise ValueError("Invalid character in binary string.")
    # Pack bits as unsigned int, then unpack as float
    packed = struct.pack('>I', bits)
    return struct.unpack('>f', packed)[0]

def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <float_number>")
        sys.exit(1)

    try:
        input_val = float(sys.argv[1])
    except ValueError:
        print("Invalid float number.")
        sys.exit(1)

    # Convert float to binary
    binary = ieee754_to_bin(input_val)
    print(f"IEEE-754 binary of {input_val:.6f}: {binary}")

    # Convert binary back to float
    recovered = bin_to_ieee754(binary)
    print(f"Recovered float from binary: {recovered:.6f}")

if __name__ == "__main__":
    main()
